src/main/java/org/opensearch/knn/plugin/script/KNNPainlessScriptUtils.java

```diff
@@ -0,0 +1,129 @@
+ * Provides late interaction scoring functionality for ColBERT-style token-level matching.
+ */
+public class KNNPainlessScriptUtils {
+
+    /**
+     * Calculates the late interaction score between query vectors and document vectors using inner product.
+     * For each query vector, finds the maximum similarity with any document vector and sums these maxima.
+     * This implements a ColBERT-style late interaction pattern for token-level matching.
+     *
+     * @param queryVectors List of query vectors
+     * @param docFieldName Name of the field in the document containing vectors
+     * @param doc Document source as a map
+     * @return Sum of maximum similarity scores
+     */
+    public static float lateInteractionScore(
+        @NonNull final List<List<Number>> queryVectors,
+        @NonNull final String docFieldName,
+        @NonNull final Map<String, Object> doc
+    ) {
+        return lateInteractionScore(queryVectors, docFieldName, doc, SpaceType.INNER_PRODUCT.getValue());
+    }
+
+    /**
+     * Calculates the late interaction score between query vectors and document vectors with specified similarity metric.
+     * For each query vector, finds the maximum similarity with any document vector and sums these maxima.
+     * This implements a ColBERT-style late interaction pattern for token-level matching.
+     *
+     * @param queryVectors List of query vectors
+     * @param docFieldName Name of the field in the document containing vectors
+     * @param doc Document source as a map
+     * @param spaceType Space type for similarity calculation: "innerproduct", "cosinesimil", "l2", "l1", "linf"
+     * @return Sum of maximum similarity scores
+     */
+    @SuppressWarnings("unchecked")
+    public static float lateInteractionScore(
+        @NonNull final List<List<Number>> queryVectors,
+        @NonNull final String docFieldName,
+        @NonNull final Map<String, Object> doc,
+        @NonNull final String spaceType
+    ) {
+        List<List<Number>> docVectors = (List<List<Number>>) doc.get(docFieldName);
+
+        if (queryVectors.isEmpty() || docVectors == null || docVectors.isEmpty()) {
+            return 0.0f;
+        }
+
+        float totalMaxSim = 0.0f;
+        SpaceType space = SpaceType.getSpace(spaceType);
+        boolean isDistanceMetric = space == SpaceType.L2 || space == SpaceType.L1 || space == SpaceType.LINF;
+
+        for (List<Number> queryVector : queryVectors) {
+            if (queryVector == null || queryVector.isEmpty()) {
+                throw new IllegalArgumentException("Every single vector within query vectors cannot be empty or null");
+            }
+
+            float[] qVec = new float[queryVector.size()];
+            for (int i = 0; i < queryVector.size(); i++) {
+                qVec[i] = queryVector.get(i).floatValue();
+            }
+
+            float maxDocTokenSim = isDistanceMetric ? Float.MAX_VALUE : Float.MIN_VALUE;
```

Comment:
I don't think you should use this. KNNScoringUtil will already be taking care of this, where large score means better result.